src/common/buffer.cc

```diff
@@ -1523,16 +1523,17 @@ static ceph::spinlock debug_lock;
    */
   char *buffer::list::c_str()
   {
-    switch (get_num_buffers()) {
-    case 0:
-      // no buffers
-      return nullptr;
-    case 1:
-      // good, we're already contiguous.
-      break;
-    default:
+    if (const auto len = length(); len == 0) {
+      return nullptr;                         // no non-empty buffers
+    } else if (len != _buffers.front().length()) {
       rebuild();
-      break;
+    } else {
+      // there two basic scenarios that hit this branch:
+      //   1. bufferlist with single bptr inside;
+      //   2. bufferlist with empty bptr at the end. splice() tries
```

Comment:
Nit: might want to be more precise here, as a single empty buffer would be handled by the "no non-empty buffers" branch.  And for the second bullet, stress that it requires a _single_ non-empty buffer.  Perhaps something like:
```suggestion
      //   1. bufferlist with single non-empty buffer;
      //   2. bufferlist with single non-empty buffer followed by empty buffer(s).
              splice() ...
```
(Multiple trailing empty buffers probably doesn't happen in practice but this code would handle it just fine.)

Reply:
> Nit: might want to be more precise here, as a single empty buffer would be handled by the "no non-empty buffers" branch.

That's the reason behind putting these words into the `else`  clause with nothing more than the comment. Sure, I will clarify this further.

> (Multiple trailing empty buffers probably doesn't happen in practice but this code would handle it just fine.)

Agreed, I don't expect them too, and thus why _slightly generalize_ is in the commit message; the crux is about handling the `splice()`-induced bptr. I think the change, while being a good improvement in the terms of readability & elegance, hides the original intention and that's the reason why I prefer to not squash.

Reply:
The updated comment reads much better, thanks!

As for squashing, the original intention is documented both in the tracker ticket and in this comment and could also be reiterated in the commit message.  I don't see any point in merging three commits for less than ten lines of code where the third commit completely overwrites the first and the second.

Reply:
Regardless, need to add `Fixes` tag.